k2pdfopt: 2.51a -> 2.53

danielfullmer opened the pull request:
###### Motivation for this change
Bump to latest version. See: https://www.willus.com/k2pdfopt/k2pdfopt_version.txt

k2pdfopt is a pain to package. Instead of shipping patches for mupdf,
leptonica, and tessertact, k2pdfopt includes modified source files for these
packages.  Individual files from the {mupdf,leptonica,tesseract}_mod/
directories are intended to replace the corresponding source files in the
upstream packages, for a particular version of that upstream package.

The previous packaging approach involved manually replacing these files and
producing patches which could then be applied to those packages.  The benefit
of creating and applying these patches in this way is that minor updates
(esp. security fixes) might still allow these patches to apply successfully.

The current approach creates these patches using nix. This has the benefit of
easier review and should hopefully be simpler to update in the future.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

danielfullmer committed:
k2pdfopt: 2.51a -> 2.53

Moredread commented on pkgs/applications/misc/k2pdfopt/default.nix:
```diff
@@ -11,44 +11,119 @@
+# easier review and should hopefully be simpler to update in the future.
+
+let
+  # Create a patch against src based on changes applied in patchCommands
```
That actually looks so useful that it might be nice to have in the standard library :)

danielfullmer replied:
Yeah, it does seem pretty generic. I wonder if there are other places in nixpkgs that could use something like this.

Moredread commented on pkgs/applications/misc/k2pdfopt/default.nix:
```diff
@@ -11,44 +11,119 @@
-    #  with more recent versions of these depencencies, and running diff.
-    mupdf_modded = mupdf.overrideAttrs (attrs: {
-      patches = attrs.patches ++ [ ./mupdf.patch ]; # Last verified with mupdf 1.16.1
+    # We use specific versions of these sources to match the versions used in the k2pdfopt source
```
If I understand it correctly, we only use this to create the patch and apply it afterwards to our own packaged mupdf with overrideAttrs.

I think it would be good to explicitly mention it here or above that the version it is applied to doesn't need to be the same.

danielfullmer replied:
That's correct. I've updated the comment here to emphasize this.

Moredread commented on pkgs/applications/misc/k2pdfopt/default.nix:
```diff
@@ -11,44 +11,119 @@
+    };
+    mupdf_modded = mupdf.overrideAttrs ({ patches ? [], ... }: {
+      patches = patches ++ [ mupdf_patch ];
+      # THis function is missing in font.c, see font-win32.c
```
THis -> This

Is this only an issue with the patch in k2pdfopt, not in uptream?

danielfullmer replied:
Thanks. Yes, this appears to be a change k2pdfopt made.

Moredread reviewed (approved):
Besides maybe improving comments in the derivation everything looks fine.

Ran nix-review and used k2pdfopt on a PDF file.

danielfullmer commented:
@Moredread Thanks for the review! I've addressed your comments, and also improved the comment paragraph explaining the unique packaging approach for k2pdfopt here.

danielfullmer committed:
k2pdfopt: set env var to find OCR language files

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/213
